myia/compile/backends/python/directed_graph.py

```diff
@@ -0,0 +1,118 @@
+"""Implementation of a directed graph.
+
+Used to convert a myia graph to a directed graph.
+Directed graph is then visited to generate Python code.
+"""
```

Comment:
Can you explain real quick why DirectedGraph exists? Why do we need another graph class? Can't we sort the nodes in Graph directly?

Reply:
Hi!

Indeed, I could simply sort the graph nodes. But I don't know initially all the graphs involved in the main graph to compile. I can find them only while recursively visiting the main graph. Sometimes, a closure `a` for a graph is only discoverable when visiting another closure `b` of same graph, because only `b` calls `a`. And, for each closure found, I need to correctly position closure code into parent graph code, so that closure is defined after parent variables it uses, and before parent variables/closures that use it.

So, to do that more conveniently, I chosed to use directed graph as strategy. It allows me to link users to used nodes (applies or closures) easily, and then visit nodes in an order when used nodes easily come before user nodes.

Goal was also to have an implementation logic more easy to understand, thus more easy to modify/debug later.

I think I could remove directed graph logic, but I would need a different way to keep trace of links between closures and applies that use them.

Reply:
Okay, let's keep it that way.